File: ppJoke/libcommon/src/main/java/com/example/libcommon/utils/AppGlobal.java
package com.example.libcommon.utils;

import android.app.Application;

import java.lang.reflect.InvocationTargetException;

/**
 * @author tengfei
 * date 2020-02-08 12:49
 * email arrayadapter.cn@gmail.com
 * description
 */
public class AppGlobal {

    private static Application sApplication;

    public static Application getApplication() {
        if (sApplication == null) {
            try {
                sApplication = (Application) Class.forName("android.app.ActivityThread")
                        .getMethod("currentApplication")
                        .invoke(null, (Object[]) null);
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            } catch (NoSuchMethodException e) {
                e.printStackTrace();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
        }
        return sApplication;
    }
}

File: ppJoke/libcommon/src/main/java/com/example/libcommon/utils/AppGlobal.kt
package com.example.libcommon.utils

import android.annotation.SuppressLint
import android.app.Application

/**
 * @author tengfei
 * date 2020-02-01 22:15
 * email arrayadapter.cn@gmail.com
 * description
 */


class AppGlobal {
    companion object {
        private var sApplication: Application? = null

        @SuppressLint("PrivateApi", "DiscouragedPrivateApi")
        fun getApp(): Application? {

            if (sApplication == null) {
                val clazz: Class<*> = Class.forName("android.app.ActivityThread")
                val method = clazz.getDeclaredMethod("currentApplication")
                sApplication = method.invoke(null) as Application?
            }

            return sApplication


        }

    }
}

